### PR TITLE
fix: forbid stores to the __debug__ builtin

### PR DESCRIPTION
**Pull request by muscar**
Addresses https://github.com/RustPython/RustPython/issues/2979.

According to [the docs][1], `__debug__` is treated as a constant by the interpreter.

This patch adds some checks to bring RustPython's behaviour in line with cpython's.

[1]: https://docs.python.org/3/library/constants.html#built-in-constants

### CODE REVIEW

**Comment by youknowone** [comment]
@HyeockJinKim could you review this PR?

File: compiler/src/compile.rs [hunk 1]
```diff
@@ -51,6 +52,15 @@ impl CallType {
     }
 }
 
+// See https://docs.python.org/3/library/constants.html#built-in-constants
+static BUILTIN_CONSTANTS: phf::Set<&'static str> = phf::phf_set! {
```
Comment (DimitrisJim) [inline compiler/src/compile.rs:56]: A constant set seems like its too much here. There's only a couple of constants (Python has 4, `None, True, False, __debug__`, and usually skips checking all of them), a slice would probably do the trick without needing an additional crate.

**Comment by DimitrisJim** [comment]
I'd also add a couple of checks in `extra_tests/snippets` if the run doesn't produce any unexpected successes. (not sure if the test suite currently checks these).

**Comment by fanninpm** [comment]
> I'd also add a couple of checks in `extra_tests/snippets` if the run doesn't produce any unexpected successes. (not sure if the test suite currently checks these).

One check is in `test_syntax.py`, but it's commented out (because doctests behave weirdly). Another check is in `test_compile.py`, which we currently don't have.

**Comment by DimitrisJim** [comment]
@fanninpm mind adding `test_compile`? would be good to test this PR against it.

**Comment by muscar** [comment]
Thanks both. `test_compile` and `test_syntax` pass.

```alex@tekton:~/src/hf/RustPython> cargo run --release -- -m test test_syntax -j 4
    Finished release [optimized] target(s) in 0.09s
     Running `target/release/rustpython -m test test_syntax -j 4`
Run tests in parallel using 4 child processes
0:00:00 load avg: 0.47 [1/1] test_syntax passed

== Tests result: SUCCESS ==

1 test OK.

Total duration: 241 ms
Tests result: SUCCESS
alex@tekton:~/src/hf/RustPython> cargo run --release -- -m test test_compile -j 4
    Finished release [optimized] target(s) in 0.09s
     Running `target/release/rustpython -m test test_compile -j 4`
Run tests in parallel using 4 child processes
0:00:01 load avg: 0.43 [1/1] test_compile passed

== Tests result: SUCCESS ==

1 test OK.

Total duration: 1 sec 156 ms
Tests result: SUCCESS
alex@tekton:~/src/hf/RustPython> cargo run --release -- -m test test_syntax -j 4
    Finished release [optimized] target(s) in 0.09s
     Running `target/release/rustpython -m test test_syntax -j 4`
Run tests in parallel using 4 child processes
0:00:00 load avg: 0.47 [1/1] test_syntax passed

== Tests result: SUCCESS ==

1 test OK.

Total duration: 241 ms
Tests result: SUCCESS
alex@tekton:~/src/hf/RustPython> cargo run --release -- -m test test_compile -j 4
    Finished release [optimized] target(s) in 0.09s
     Running `target/release/rustpython -m test test_compile -j 4`
Run tests in parallel using 4 child processes
0:00:01 load avg: 0.43 [1/1] test_compile passed

== Tests result: SUCCESS ==

1 test OK.

Total duration: 1 sec 156 ms
Tests result: SUCCESS
```

File: compiler/src/compile.rs [hunk 2]
```diff
@@ -51,6 +52,13 @@ impl CallType {
     }
 }
 
+// See https://docs.python.org/3/library/constants.html#built-in-constants
+static BUILTIN_CONSTANTS: [&str; 1] = ["__debug__"];
+
+fn is_builtin_constant(name: &str) -> bool {
+    BUILTIN_CONSTANTS.contains(&name)
+}
```
Comment (youknowone) [inline compiler/src/compile.rs:60]:
```suggestion
fn is_builtin_constant(name: &str) -> bool {
    // See https://docs.python.org/3/library/constants.html#built-in-constants
    const BUILTIN_CONSTANTS: &[&str] = &["__debug__"];

    BUILTIN_CONSTANTS.contains(&name)
}
```

**Review by youknowone — DISMISSED** [review]

**Comment by DimitrisJim** [comment]
There's still a couple of things missing (see `test_debug_assignment` in `test_compile.py` and some of the doctests in `test_syntax`).

Additionally, supplying `__debug__` as a kwarg is still not handled (`def foo(**_): pass; foo(__debug__='yup')`). I say we merge this now since its in the right direction (and only needs extending to handle these other cases), I'll post this comment to the issue so you (or anyone else) can try and tackle the remaining cases.

**Review by fanninpm — CHANGES_REQUESTED** [review]

File: Lib/test/test_compile.py [hunk 3]
```diff
@@ -61,7 +61,6 @@ def test_none_keyword_arg(self):
     def test_duplicate_global_local(self):
         self.assertRaises(SyntaxError, exec, 'def f(a): global a; a = 1')
 
-    # TODO: RUSTPYTHON
     @unittest.expectedFailure
```
[the remark this answers is not in the document]
Comment (DimitrisJim) [reply Lib/test/test_compile.py:65]: huh, guess not or else the CI would fail with the unexpected success.

**Comment by muscar** [comment]
@DimitrisJim which part of `test_debug_assignment`? Using `setattr` to se the value of `__debug__` works in python:
```python
Python 3.9.7 (default, Sep  3 2021, 12:37:55)
[Clang 12.0.5 (clang-1205.0.22.9)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import builtins
>>> setattr(builtins, '__debug__', 5)
>>> builtins.__debug__
5
```

And which of the cases in `test_syntax` do you mean?

I'll have a look at using `__debug__` as a kwarg when calling a function, which doesn't seem to be covered by the PR.

**Comment by DimitrisJim** [comment]
> Using setattr to se the value of `__debug__` works in python:

The `setattr` works but the value of `__debug__` isn't changed (i.e not the value of `builtins.__debug__`, that _is_ changed).

```python
setattr(builtins, "__debug__", "whatever")
__debug__  # still is True
```

This might be more involved as a change though.

> And which of the cases in test_syntax do you mean

It's a couple of the docstrings, you'll need to uncomment the line at the end of `test_syntax` that runs them in order to see which of them fail.

**Comment by muscar** [comment]
@DimitrisJim updated so that the `setattr` on `__debug__` works as in cpython. Also uncommented the relevant tests in `test_compile`. Still not running the docstrings in `test_syntax`, but I _think_ that the extra cases should be covered:

```python
Welcome to the magnificent Rust Python 0.1.2 interpreter 😱 🖖
>>>>> __debug__ = 1
SyntaxError: cannot assign to __debug__ at line 1 column 13
__debug__ = 1
            ↑
>>>>> __debug__ += 1
SyntaxError: cannot assign to __debug__ at line 1 column 14
__debug__ += 1
             ↑
>>>>> def foo(__debug__): pass
SyntaxError: cannot assign to __debug__ at line 1 column 1
def foo(__debug__): pass
↑
>>>>> def foo(__debug__=1): pass
SyntaxError: cannot assign to __debug__ at line 1 column 19
def foo(__debug__=1): pass
                  ↑
>>>>> def foo(**kwargs): pass
>>>>> foo(__debug__=1)
SyntaxError: cannot assign to __debug__ at line 1 column 1
foo(__debug__=1)
↑
>>>>> import builtins
>>>>> builtins.__debug__ = 1
SyntaxError: cannot assign to __debug__ at line 1 column 22
builtins.__debug__ = 1
                     ↑
>>>>> setattr(builtins, '__debug__', 'hi!')
>>>>> __debug__
True
```

File: vm/src/builtins/object.rs [hunk 4]
```diff
@@ -323,6 +323,19 @@ pub(crate) fn setattr(
 ) -> PyResult<()> {
     vm_trace!("object.__setattr__({:?}, {}, {:?})", obj, attr_name, value);
 
+    if attr_name.as_str() == "__debug__" {
```
Comment (DimitrisJim) [inline vm/src/builtins/object.rs:326]: I'd revert these changes. This adds a check for every single look-up and doesn't match behavior expected (the `setattr` is supposed to succeed in changing the value). 

Looking at the disassembly of looking up `__debug__` it seems as if `True` is immediately loaded, indicating no dynamic look-up is performed for `__debug__`. CPython probably detects `__debug__` being looked-up in their peephole optimizer and replaces it with `True` (that's why I said this change is probably a bit involved).

**Review by DimitrisJim — CHANGES_REQUESTED** [review]

File: vm/src/builtins/object.rs (hunk 4 above)
Comment (muscar) [reply vm/src/builtins/object.rs]: Fair enough. Better left for another PR.

File: compiler/src/compile.rs [hunk 5]
```diff
@@ -375,26 +383,39 @@ impl Compiler {
...
+                name
+            ))));
+        }
+        if usage == NameUsage::Delete && is_forbidden_name(name) {
+            return Err(self.error(CompileErrorType::SyntaxError(format!(
+                "cannot delete {}",
+                name
+            ))));
+        }
+        Ok(())
```
Comment (youknowone) [inline compiler/src/compile.rs:411]:
```suggestion
        if usage == NameUsage::Store && is_forbidden_name(name) {
            return Err(self.error(CompileErrorType::SyntaxError(format!(
                "cannot assign to {}",
                name
            ))));
        }
        if usage == NameUsage::Delete && is_forbidden_name(name) {
            return Err(self.error(CompileErrorType::SyntaxError(format!(
                "cannot delete {}",
                name
            ))));
        }
        Ok(())
```

Comment (youknowone) [reply compiler/src/compile.rs]: @muscar oh, I thought I changed it quickly enough but it wasn't. could you check the newer version of this comment? I am sorry to make confusion.

**Comment by DimitrisJim** [comment]
`test_debug_assignment` should be skipped again (as it previously was), that's causing the CI failure.

**Review by DimitrisJim — APPROVED** [review]

**Pull request merged**